File: yggdrasil/examples/hello/src/hello.cpp
#include "YggInterface.hpp"
#include <string>
#include <iostream>
using namespace std;

#define BSIZE 512 // the max

int main(int argc, char *argv[]) {
  int ret = 1;
  char buf[BSIZE];

  cout << "Hello from C++\n";
  
  /* Matching with the the model yaml */
  YggInput inf("inFile"); 
  YggOutput outf("outFile");
  YggInput inq("helloQueueIn");
  YggOutput outq("helloQueueOut");
  cout << "hello(CPP): Created I/O channels\n";

  // Receive input from the local file
  ret = inf.recv(buf, BSIZE);
  if (ret < 0) {
    printf("hello(CPP): ERROR FILE RECV\n");
    return -1;
  }
  int bufsiz = ret;
  printf("hello(CPP): Received %d bytes from file: %s\n", bufsiz, buf);

  // Send output to queue
  ret = outq.send(buf, bufsiz);
  if (ret != 0) {
    printf("hello(CPP): ERROR QUEUE SEND\n");
    return -1;
  }
  printf("hello(CPP): Sent to outq\n");

  // Receive input from queue
  ret = inq.recv(buf, BSIZE);
  if (ret < 0) {
    printf("hello(CPP): ERROR QUEUE RECV\n");
    return -1;
  }
  bufsiz = ret;
  printf("hello(CPP): Received %d bytes from queue: %s\n", bufsiz, buf);

  // Send output to local file
  ret = outf.send(buf, bufsiz);
  if (ret != 0) {
    printf("hello(CPP): ERROR FILE SEND\n");
    return -1;
  }
  printf("hello(CPP): Sent to outf\n");

  cout << "Goodbye from C++\n";
  return 0;
}

File: yggdrasil/examples/hello/src/hello.f90
PROGRAM hello
  USE fygg

  logical :: ret = .true.
  integer :: bufsiz
  character(len = 512) :: buf = ""
  TYPE(yggcomm) :: inf, outf, inq, outq

  PRINT *, "Hello from Fortran"

  ! Ins/outs matching with the the model yaml
  inf = ygg_input("inFile")
  outf = ygg_output("outFile")
  inq = ygg_input("helloQueueIn")
  outq = ygg_output("helloQueueOut")
  PRINT *, "hello(Fortran): Created I/O channels"

  ! Receive input from a local file
  bufsiz = len(buf)
  ret = ygg_recv(inf, buf, bufsiz)
  IF (.not.ret) THEN
     PRINT *, "hello(Fortran): ERROR FILE RECV"
     STOP 1
  END IF
  PRINT *, "hello(Fortran): Received ", bufsiz, &
       "bytes from file: ", buf

  ! Send output to the output queue
  ret = ygg_send(outq, buf, bufsiz)
  IF (.not.ret) THEN
     PRINT *, "hello(Fortran): ERROR QUEUE SEND"
     STOP 1
  END IF
  PRINT *, "hello(Fortran): Sent to outq"
  
  ! Receive input form the input queue
  bufsiz = len(buf)
  ret = ygg_recv(inq, buf, bufsiz)
  IF (.not.ret) THEN
     PRINT *, "hello(Fortran): ERROR QUEUE RECV"
     STOP 1
  END IF
  PRINT *, "hello(Fortran): Received ", bufsiz, &
       "bytes from queue: ", buf

  ! Send output to a local file
  ret = ygg_send(outf, buf, bufsiz)
  IF (.not.ret) THEN
     PRINT *, "hello(Fortran): ERROR FILE SEND"
     STOP 1
  END IF
  PRINT *, "hello(Fortran): Sent to outf"

  PRINT *, "Goodbye from Fortran"
  
END PROGRAM hello
